consumer/high_level_pull_consumer.go

```diff
@@ -0,0 +1,481 @@
+
+		if hpc.model == Clustering {
+			// set retry topic
+			retryTopic := internal.GetRetryTopic(hpc.consumerGroup)
+			sub := buildSubscriptionData(retryTopic, MessageSelector{TAG, _SubAll})
+			hpc.subscriptionDataTable.Store(retryTopic, sub)
+		}
+
+		if hpc.model == Clustering {
+			hpc.option.ChangeInstanceNameToPID()
+			hpc.storage = NewRemoteOffsetStore(hpc.consumerGroup, hpc.client, hpc.namesrv)
+		} else {
+			hpc.storage = NewLocalFileOffsetStore(hpc.consumerGroup, hpc.client.ClientID())
+		}
+
+		hpc.client.Start()
+		atomic.StoreInt32(&hpc.state, int32(internal.StateRunning))
+		hpc.consumerStartTimestamp = time.Now().UnixNano() / int64(time.Millisecond)
+		if err != nil {
+			return
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	hpc.client.UpdateTopicRouteInfo()
+	for k := range hpc.subscribedTopic {
+		_, exist := hpc.topicSubscribeInfoTable.Load(k)
+		if !exist {
+			hpc.client.Shutdown()
+			return fmt.Errorf("the topic=%s route info not found, it may not exist", k)
+		}
+	}
+	hpc.client.SendHeartbeatToAllBrokerWithLock()
+	hpc.Rebalance()
+	return err
+}
+
+func (hpc *defaultHighLevelPullConsumer) Shutdown() error {
+	var err error
+	hpc.closeOnce.Do(func() {
+		close(hpc.done)
+		hpc.UnregisterConsumer(hpc.consumerGroup)
+		atomic.StoreInt32(&hpc.state, int32(internal.StateShutdown))
+		mqs := make([]*primitive.MessageQueue, 0)
+		hpc.processQueueTable.Range(func(key, value interface{}) bool {
+			k := key.(primitive.MessageQueue)
+			pq := value.(*processQueue)
+			pq.WithDropped(true)
+			// close msg channel using RWMutex to make sure no data was writing
+			pq.mutex.Lock()
+			close(pq.msgCh)
+			pq.mutex.Unlock()
+			mqs = append(mqs, &k)
+			return true
+		})
+		hpc.storage.persist(mqs)
+		hpc.client.Shutdown()
+	})
+	return err
+}
+
+func (hpc *defaultHighLevelPullConsumer) Subscribe(topic string, selector MessageSelector) error {
+	if atomic.LoadInt32(&hpc.state) == int32(internal.StateStartFailed) ||
+		atomic.LoadInt32(&hpc.state) == int32(internal.StateShutdown) {
+		return errors2.ErrStartTopic
+	}
+
+	if "" == topic {
+		rlog.Error(fmt.Sprintf("topic is null"), nil)
+	}
+
+	if hpc.option.Namespace != "" {
+		topic = hpc.option.Namespace + "%" + topic
+	}
+	data := buildSubscriptionData(topic, selector)
+	hpc.subscriptionDataTable.Store(topic, data)
+	hpc.subscribedTopic[topic] = ""
+	routeData, _, err := hpc.namesrv.UpdateTopicRouteInfo(topic)
+	if err == nil {
+		hpc.topicSubscribeInfoTable.Store(topic, routeData)
+	}
+	hpc.Rebalance()
+	return nil
+}
+
+func (hpc *defaultHighLevelPullConsumer) Unsubscribe(topic string) error {
+	if hpc.option.Namespace != "" {
+		topic = hpc.option.Namespace + "%" + topic
+	}
+
+	if "" == topic {
+		rlog.Error(fmt.Sprintf("topic is null"), nil)
+	}
+
+	hpc.subscriptionDataTable.Delete(topic)
+	return nil
+}
+
+func (hpc *defaultHighLevelPullConsumer) Pull(ctx context.Context, topic string, numbers int) (*primitive.PullResult, error) {
+	queue, err := hpc.getNextQueueOf(topic)
```

Comment:
send a message to the topic can not be consumed immediatly, If the topic have 10 queues, we have to wait for 180 seconds, because no message is in other 9 queues, and each queue will wait for _BrokerSuspendMaxTime(20 seconds)